gflownet/utils/buffer.py

```diff
@@ -143,22 +148,29 @@ def _add_greater(
         rewards,
         energies,
         it,
+        allow_duplicate_states=False,
     ):
-        rewards_old = self.replay["reward"].values
-        rewards_new = rewards.copy()
-        while np.max(rewards_new) > np.min(rewards_old):
-            idx_new_max = np.argmax(rewards_new)
-            readable_state = self.env.state2readable(states[idx_new_max])
-            if not self.replay["state"].isin([readable_state]).any():
-                self.replay.iloc[self.replay.reward.argmin()] = {
-                    "state": self.env.state2readable(states[idx_new_max]),
-                    "traj": self.env.traj2readable(trajs[idx_new_max]),
-                    "reward": rewards[idx_new_max],
-                    "energy": energies[idx_new_max],
+        for idx, (state, traj, reward, energy) in enumerate(
+            zip(states, trajs, rewards, energies)
+        ):
+            if allow_duplicate_states is False and state in self.replay_states.values():
```

Comment:
This doesn't seem to be working for tensor states (in particular for trees): it results in `RuntimeError: Boolean value of Tensor with more than one value is ambiguous`.

Reply:
Yes, I am aware. It is mentioned in the description, in "Other notes":
> Backward sampling does not work yet with the tetris environment because the states are tensors and I need to adapt the code.

I am sorry but all these changes took me too long and I wanted to give priority to merging sooner than later. It's not that this PR breaks it. It didn't work before either any way. And I am planning a big refactor of the buffer class, so I thought I'll fix that later.